circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/AnsweringNavigator.kt

```diff
@@ -0,0 +1,136 @@
+  block: suspend CoroutineScope.(result: T) -> Unit,
+): GoToNavigator {
+  val backStack = LocalBackStack.current ?: return fallbackNavigator
+  return rememberAnsweringNavigator(backStack, resultType, block)
+}
+
+/**
+ * A reified version of [rememberAnsweringNavigator]. See documented overloads of this function for
+ * more information.
+ */
+@Composable
+public inline fun <reified T : PopResult> rememberAnsweringNavigator(
+  backStack: BackStack<out BackStack.Record>,
+  noinline block: suspend CoroutineScope.(result: T) -> Unit,
+): GoToNavigator {
+  return rememberAnsweringNavigator(backStack, T::class, block)
+}
+
+/**
+ * Returns a [GoToNavigator] that answers with the given [resultType] using the given [backStack].
+ *
+ * Handling of the result type [T] should be handled in the [block] parameter and is guaranteed to
+ * only be called _at most_ once. It may never be called if the navigated screen never pops with a
+ * result (of equivalent type) back.
+ *
+ * Note that [resultType] is a simple instance check, so subtypes may also be valid answers.
+ *
+ * ## Example
+ *
+ * ```kotlin
+ * val pickPhotoNavigator = rememberAnsweringNavigator(backStack, PickPhotoScreen.Result::class) { result: PickPhotoScreen.Result ->
+ *   // Do something with the result!
+ * }
+ *
+ * return State(...) { event ->
+ *   when (event) {
+ *     is PickPhoto -> pickPhotoNavigator.goTo(PickPhotoScreen)
+ *   }
+ * }
+ *
+ * // In PickPhotoScreen
+ * navigator.pop(PickPhotoScreen.Result(...))
+ * ```
+ */
+@Composable
+public fun <T : PopResult> rememberAnsweringNavigator(
+  backStack: BackStack<out BackStack.Record>,
+  resultType: KClass<T>,
+  block: suspend CoroutineScope.(result: T) -> Unit,
+): GoToNavigator {
+  val currentBackStack by rememberUpdatedState(backStack)
+  val currentResultType by rememberUpdatedState(resultType)
+
+  // Top screen at the start, so we can ensure we only collect the result if
+  // we've returned to this screen
+  val initialRecordKey = rememberSaveable {
+    currentBackStack.topRecord?.key ?: error("Navigator must have a top screen at start.")
+  }
+
+  // Key for the resultKey, so we can track who owns this requested result
+  val key = rememberSaveable { uuid4().toString() }
+
+  // Current top record of the navigator
+  val currentTopRecordKey by remember { derivedStateOf { currentBackStack.topRecord!!.key } }
+
+  // Track whether we've actually gone to the next record yet
+  var launched by rememberSaveable { mutableStateOf(false) }
+
+  // Collect the result if we've launched and now returned to the initial record
+  if (launched && currentTopRecordKey == initialRecordKey) {
+    LaunchedEffect(key) {
+      println("⚡️${initialRecordKey.take(8)} is awaiting result with $key")
+      val result = currentBackStack.topRecord!!.awaitResult(key) ?: return@LaunchedEffect
+      launched = false
+      if (currentResultType.isInstance(result)) {
+        println("✅ Answering with result: $result")
```

Comment:
@ZacSweers we want to clean up the logging?

Reply:
Done! [2ff6ceb](https://github.com/slackhq/circuit/pull/1154/commits/2ff6cebdeecd6fee4fb1cd0bb7c5402a9ec897e2)

Keeping the test logging for now as I think there's some use we can get out of it in followups